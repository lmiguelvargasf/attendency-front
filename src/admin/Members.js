import React from 'react'
import useAxios from 'axios-hooks'
import { Table, Space } from 'antd'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faEdit, faTrash } from '@fortawesome/free-solid-svg-icons'

const Members = () => {
  const [{ data: members, loading, error }] = useAxios(
    `${process.env.REACT_APP_API_URL}/members/`
  )

  if (loading) return <p data-testid='loading'>Loading...</p>
  if (error) return <p data-testid='error'>Error!</p>

  const columns = [
    {
      title: 'First Name',
      dataIndex: 'firstName',
      key: 'firstName',
      render: firstName => <a>{firstName}</a>
    },
    {
      title: 'Last Name',
      dataIndex: 'lastName',
      key: 'lastName'
    },
    {
      title: 'Email',
      dataIndex: 'email',
      key: 'email'
    },
    {
      title: 'Action',
      key: 'action',
      render: (text, record) => (
        <Space size='middle'>
          <FontAwesomeIcon icon={faEdit} />
          <FontAwesomeIcon icon={faTrash} />
        </Space>
      )
    }
  ]

  return (
    <Table
      data-testid='member-table'
      columns={columns}
      dataSource={members}
      pagination={false}
    />
  )
}

export default Members
