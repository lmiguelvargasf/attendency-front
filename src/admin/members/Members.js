import React, { useEffect, useReducer } from 'react'
import { Route, Switch } from 'react-router-dom'
import useAxios from 'axios-hooks'
import MembersTable from './MembersTable'
import CreateMember from './CreateMember'
import EditMeeting from './EditMember'
import memberReducer from './memberReducer'

const Members = () => {
  const [members, dispatch] = useReducer(memberReducer, [])
  const [{ data, loading, error }] = useAxios(`${process.env.REACT_APP_API_URL}/members/`)
  useEffect(() => { dispatch({ type: 'LOAD', members: data }) }, [data])

  if (loading) return <p data-testid='loading'>Loading...</p>
  if (error) return <p data-testid='error'>Error!</p>

  return (
    <Switch>
      <Route
        path='/admin/members/create'
        render={props => <CreateMember {...props} addMember={member => dispatch({ type: 'ADD', member })} />}
      />
      <Route
        path='/admin/members/:key/edit'
        render={props => <EditMeeting {...props} updateMembers={(member, index) => dispatch({ type: 'UPDATE', member, index })} />}
      />
      <Route
        path='/admin/members'
        render={props => <MembersTable {...props} members={members} removeMember={member => dispatch({ type: 'REMOVE', member })} />}
      />
    </Switch>
  )
}

export default Members
