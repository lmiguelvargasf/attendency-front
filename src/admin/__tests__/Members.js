import React from 'react'
import { render, cleanup } from '@testing-library/react'
import Members from '../Members'
import useAxios from 'axios-hooks'
jest.mock('axios-hooks')

const BASE_API_URL = process.env.REACT_APP_API_URL
const TABLE_TEST_ID = 'member-table'
const fakeData = [
  {
    key: 1,
    url: `${BASE_API_URL}/members/1`,
    firstName: 'Isaac',
    lastName: 'Newton',
    email: 'isaac.newton@gmail.com'
  },
  {
    key: 2,
    url: `${BASE_API_URL}/members/2`,
    firstName: 'Nikola',
    lastName: 'Tesla',
    email: 'nikola.tesla@gmail.com'
  }
]

describe('Members component', () => {
  afterEach(cleanup)

  describe('loading data', () => {
    beforeAll(() => {
      useAxios.mockImplementation(url => [{
        data: null,
        loading: true,
        error: null
      }])
    })

    it('matches snapshot when loading data', () => {
      const { asFragment } = render(<Members />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders loading', () => {
      const { getByTestId, queryByTestId } = render(<Members />)
      expect(queryByTestId(TABLE_TEST_ID)).toBeNull()
      expect(getByTestId('loading')).not.toBeNull()
      expect(queryByTestId('error')).toBeNull()
    })
  })

  describe('error', () => {
    beforeAll(() => {
      useAxios.mockImplementation(url => [{
        data: null,
        loading: false,
        error: 'Error'
      }])
    })

    it('matches snapshot when having error', () => {
      const { asFragment } = render(<Members />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders error', () => {
      const { getByTestId, queryByTestId } = render(<Members />)
      expect(queryByTestId(TABLE_TEST_ID)).toBeNull()
      expect(queryByTestId('loading')).toBeNull()
      expect(getByTestId('error')).not.toBeNull()
    })
  })

  describe('members table', () => {
    beforeAll(() => {
      useAxios.mockImplementation(url => [{
        data: fakeData,
        loading: false,
        error: null
      }])
    })

    it('matches snapshop when displaying table', () => {
      const { asFragment } = render(<Members />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('renders members table', async () => {
      useAxios.mockImplementation(url => [{
        data: fakeData,
        loading: false,
        error: null
      }])
      const { getByTestId, queryByTestId } = render(<Members />)
      expect(getByTestId(TABLE_TEST_ID)).not.toBeNull()
      expect(queryByTestId('loading')).toBeNull()
      expect(queryByTestId('error')).toBeNull()
    })
  })
})
