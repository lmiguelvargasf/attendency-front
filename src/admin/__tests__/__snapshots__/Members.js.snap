// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Members component error matches snapshot when having error 1`] = `
<DocumentFragment>
  <p
    data-testid="error"
  >
    Error!
  </p>
</DocumentFragment>
`;

exports[`Members component loading data matches snapshot when loading data 1`] = `
<DocumentFragment>
  <p
    data-testid="loading"
  >
    Loading...
  </p>
</DocumentFragment>
`;

exports[`Members component members table matches snapshop when displaying table 1`] = `
<DocumentFragment>
  <div
    class="ant-table-wrapper"
  >
    <div
      class="ant-spin-nested-loading"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table"
          data-testid="member-table"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                    >
                      First Name
                    </th>
                    <th
                      class="ant-table-cell"
                    >
                      Last Name
                    </th>
                    <th
                      class="ant-table-cell"
                    >
                      Email
                    </th>
                    <th
                      class="ant-table-cell"
                    >
                      Action
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="1"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <a>
                        Isaac
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      Newton
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      isaac.newton@gmail.com
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="ant-space ant-space-horizontal"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 16px;"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-edit fa-w-18 "
                            data-icon="edit"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 576 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                        <div
                          class="ant-space-item"
                          style="margin-right: 16px;"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-trash fa-w-14 "
                            data-icon="trash"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 448 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-times fa-w-11 "
                            data-icon="times"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 352 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0"
                    data-row-key="2"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <a>
                        Nikola
                      </a>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      Tesla
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      nikola.tesla@gmail.com
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="ant-space ant-space-horizontal"
                      >
                        <div
                          class="ant-space-item"
                          style="margin-right: 16px;"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-edit fa-w-18 "
                            data-icon="edit"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 576 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                        <div
                          class="ant-space-item"
                          style="margin-right: 16px;"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-trash fa-w-14 "
                            data-icon="trash"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 448 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                        <div
                          class="ant-space-item"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-times fa-w-11 "
                            data-icon="times"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 352 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
